File: tests/test_laplacian.py
"""Tests for the cotangent Laplacian and lumped mass matrix."""

import numpy as np
import scipy.sparse as sp
import pytest

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from mesh_parameterization.laplacian import cotangent_laplacian


# ---------------------------------------------------------------------------
# Fixtures
# ---------------------------------------------------------------------------

def single_triangle():
    """One equilateral triangle."""
    s = np.sqrt(3) / 2
    verts = np.array([[0.0, 0.0, 0.0],
                      [1.0, 0.0, 0.0],
                      [0.5, s,  0.0]])
    faces = np.array([[0, 1, 2]])
    return verts, faces


def unit_square_mesh():
    """Unit square split into two triangles.

    Vertices:  0=(0,0,0)  1=(1,0,0)  2=(1,1,0)  3=(0,1,0)
    Faces:     [0,1,2]  [0,2,3]
    """
    verts = np.array([[0.0, 0.0, 0.0],
                      [1.0, 0.0, 0.0],
                      [1.0, 1.0, 0.0],
                      [0.0, 1.0, 0.0]])
    faces = np.array([[0, 1, 2],
                      [0, 2, 3]])
    return verts, faces


def regular_icosphere(subdivisions=1):
    """Small closed triangle mesh (icosphere)."""
    import trimesh
    mesh = trimesh.creation.icosphere(subdivisions=subdivisions)
    return np.array(mesh.vertices), np.array(mesh.faces)


# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

class TestCotangentLaplacian:

    def test_output_shapes(self):
        v, f = unit_square_mesh()
        L, M, cw = cotangent_laplacian(v, f)
        n = len(v)
        assert L.shape == (n, n)
        assert M.shape == (n, n)
        assert cw.shape == (len(f), 3)

    def test_L_is_sparse(self):
        v, f = unit_square_mesh()
        L, M, _ = cotangent_laplacian(v, f)
        assert sp.issparse(L)
        assert sp.issparse(M)

    def test_L_symmetric(self):
        v, f = unit_square_mesh()
        L, _, _ = cotangent_laplacian(v, f)
        diff = (L - L.T).toarray()
        assert np.allclose(diff, 0, atol=1e-10)

    def test_L_row_sums_zero(self):
        """Laplacian rows must sum to zero (constant functions are null)."""
        v, f = regular_icosphere()
        L, _, _ = cotangent_laplacian(v, f)
        row_sums = np.array(L.sum(axis=1)).ravel()
        assert np.allclose(row_sums, 0, atol=1e-8)

    def test_L_negative_semi_definite(self):
        """All eigenvalues of L must be ≤ 0."""
        v, f = regular_icosphere()
        L, _, _ = cotangent_laplacian(v, f)
        # Check via diagonal: should be non-positive
        diag = L.diagonal()
        assert np.all(diag <= 1e-10)

    def test_M_diagonal_positive(self):
        v, f = unit_square_mesh()
        _, M, _ = cotangent_laplacian(v, f)
        diag = M.diagonal()
        assert np.all(diag > 0)

    def test_M_sums_to_total_area(self):
        """Sum of lumped masses must equal total mesh area."""
        v, f = unit_square_mesh()
        _, M, _ = cotangent_laplacian(v, f)
        total_area = 1.0   # unit square
        assert np.isclose(M.diagonal().sum(), total_area, rtol=1e-6)

    def test_equilateral_triangle_cotangents(self):
        """All angles 60° → cot = 1/√3 ≈ 0.5774."""
        v, f = single_triangle()
        _, _, cw = cotangent_laplacian(v, f)
        expected = 1.0 / np.sqrt(3)
        assert np.allclose(cw[0], expected, atol=1e-6)

    def test_right_triangle_cotangents(self):
        """Right-angle triangle at v0: cot(90°)=0, cot(45°)=1."""
        verts = np.array([[0.0, 0.0, 0.0],
                          [1.0, 0.0, 0.0],
                          [0.0, 1.0, 0.0]])
        faces = np.array([[0, 1, 2]])
        _, _, cw = cotangent_laplacian(verts, faces)
        assert np.isclose(cw[0, 0], 0.0, atol=1e-6)   # right angle at v0
        assert np.isclose(cw[0, 1], 1.0, atol=1e-6)   # 45° at v1
        assert np.isclose(cw[0, 2], 1.0, atol=1e-6)   # 45° at v2

    def test_degenerate_triangle_no_crash(self):
        """Flat (zero-area) triangle should not raise an exception."""
        verts = np.array([[0.0, 0.0, 0.0],
                          [1.0, 0.0, 0.0],
                          [2.0, 0.0, 0.0]])
        faces = np.array([[0, 1, 2]])
        L, M, cw = cotangent_laplacian(verts, faces)
        assert L.shape == (3, 3)

File: tests/test_laplacian.cpp
#include <gtest/gtest.h>
#include "meshparam/laplacian.h"
#include <Eigen/Core>

namespace {

// Simple test mesh: a single quad split into 2 triangles
// Vertices form a unit square in the XY plane:
//   v2---v3
//   |  / |
//   v0---v1
void make_unit_square(Eigen::MatrixXd& V, Eigen::MatrixXi& F) {
    V.resize(4, 3);
    V << 0, 0, 0,
         1, 0, 0,
         0, 1, 0,
         1, 1, 0;

    F.resize(2, 3);
    F << 0, 1, 2,
         1, 3, 2;
}

// Equilateral triangle
void make_equilateral(Eigen::MatrixXd& V, Eigen::MatrixXi& F) {
    V.resize(3, 3);
    V << 0, 0, 0,
         1, 0, 0,
         0.5, std::sqrt(3.0)/2.0, 0;

    F.resize(1, 3);
    F << 0, 1, 2;
}

} // anonymous namespace

TEST(Laplacian, CotangentMatrixSize) {
    Eigen::MatrixXd V;
    Eigen::MatrixXi F;
    make_unit_square(V, F);

    auto L = meshparam::cotangent_laplacian(V, F);
    EXPECT_EQ(L.rows(), 4);
    EXPECT_EQ(L.cols(), 4);
}

TEST(Laplacian, CotangentMatrixSymmetric) {
    Eigen::MatrixXd V;
    Eigen::MatrixXi F;
    make_unit_square(V, F);

    auto L = meshparam::cotangent_laplacian(V, F);

    // Check symmetry
    for (int i = 0; i < L.rows(); ++i) {
        for (int j = 0; j < L.cols(); ++j) {
            EXPECT_NEAR(L.coeff(i, j), L.coeff(j, i), 1e-12);
        }
    }
}

TEST(Laplacian, RowSumsZero) {
    Eigen::MatrixXd V;
    Eigen::MatrixXi F;
    make_unit_square(V, F);

    auto L = meshparam::cotangent_laplacian(V, F);

    // Row sums should be approximately zero (L * ones = 0)
    Eigen::VectorXd ones = Eigen::VectorXd::Ones(L.rows());
    Eigen::VectorXd result = L * ones;
    for (int i = 0; i < result.size(); ++i) {
        EXPECT_NEAR(result(i), 0.0, 1e-10);
    }
}

TEST(Laplacian, MassMatrixSize) {
    Eigen::MatrixXd V;
    Eigen::MatrixXi F;
    make_unit_square(V, F);

    auto B = meshparam::mass_matrix(V, F);
    EXPECT_EQ(B.rows(), 4);
    EXPECT_EQ(B.cols(), 4);
}

TEST(Laplacian, MassMatrixPositive) {
    Eigen::MatrixXd V;
    Eigen::MatrixXi F;
    make_unit_square(V, F);

    auto B = meshparam::mass_matrix(V, F);

    // Diagonal entries should be positive
    for (int i = 0; i < B.rows(); ++i) {
        EXPECT_GT(B.coeff(i, i), 0.0);
    }
}

TEST(Laplacian, MassMatrixAreaConsistency) {
    Eigen::MatrixXd V;
    Eigen::MatrixXi F;
    make_equilateral(V, F);

    auto B = meshparam::mass_matrix(V, F);

    // Sum of diagonal = total area for Voronoi mass matrix
    double total_mass = 0;
    for (int i = 0; i < B.rows(); ++i) {
        total_mass += B.coeff(i, i);
    }
    double expected_area = std::sqrt(3.0) / 4.0;  // area of unit equilateral triangle
    EXPECT_NEAR(total_mass, expected_area, 1e-10);
}
